### ISSUE TITLE
_per_thread() data in separate compilation units

### ISSUE THREAD

**Issue by tjhei** [taskflow/taskflow#245]
When an tf::Executor is created in one translation unit and then used in another translation unit (or shared library?), the per-thread data inside the Executor is invalid. This causes asserts like

```
../bundled/taskflow-2.7.0/include/taskflow/core/executor.hpp:976: void tf::Executor::_invoke_dynamic_work_external(tf::Node*, tf::Graph&, bool): Assertion `worker &&
worker->executor == this' failed.
```
to trigger.

The problem is
https://github.com/taskflow/taskflow/blob/9d17ee3fb28ef8b9b92cd28d10d7ac840ad33de8/taskflow/core/executor.hpp#L446
which will create a different copy in each translation unit, also see
https://stackoverflow.com/questions/185624/static-variables-in-an-inlined-function
for more details.

I am attaching an example that uses a shared library for this. Compile and run with:
```
export LD_LIBRARY_PATH=`pwd`
# 1. make libfoo.so
g++ --std=c++14 -Wall -I ~/taskflow/installed/include/ -O2 -pthread -fpic -c other.cc
g++ -shared -o libfoo.so other.o

# 2. compile a.out and link libfoo.so
g++ --std=c++14 -Wall -I ~/taskflow/installed/include/ -O2 -pthread -o a.out -L. main.cc -lfoo

# 3. run
./a.out
```

source code to reproduce:
[taskflow-thread-local-bug.zip](https://github.com/taskflow/taskflow/files/5514222/taskflow-thread-local-bug.zip)

**Comment by tsung-wei-huang** [comment]
Hi, @tjhei thank you for the issue. Indeed, there is a similar issue #158 about this initialization order of static variables used in header-only library. I am solving this problem in the next major release 3.0 through `inline variable` in C++17. As far as I know, there is no easy hack to get around with this issue without C++17 `inline variable`. Does your system support C++17? If yes, would you please try the master branch with this example and see if the problem still exists? 

I tested it on my side on the master branch using `g++ 8, 9, 10` with `-std=c++17` and no errors occur.

**Comment by tjhei** [comment]
> Does your system support C++17?

We currently require only c++14 in deal.II, so this is not ideal.

**Comment by tsung-wei-huang** [comment]
Another workaround is to maintain a custom mapping between the worker thread to its id, instead of using `tf::Executor::this_worker_id`. Do you think this is possible in your engine?

**Comment by tjhei** [comment]
If I understand you correctly, the options are:
a) switch to 3.x (requires c++17)
b) add a map thread_id -> worker_id inside the Executor
c) make 2.x a shared library and add a static thread_local member for the worker_id

Is this correct?

Would you be willing to provide b) or c) in a 2.x release? Or would you consider making the c++17 requirement optional with b) or c) as a workaround (assuming this is the only c++17 feature you need at this point)? I can try to help if you decide where you want to go with this.

Thanks!

**Comment by tsung-wei-huang** [comment]
@tjhei I would definitely prefer a) and then c). Providing a shared library is also one thing I would like to add on top of taskflow and it can be part of 2.x, 3.x, and other future releases. I think (b) may be the fastest but it is not the best solution. Maybe we should look at (c) for now?

**Comment by bangerth** [comment]
The solutions used in libraries that could not use `thread_local` variables was to use
```
  static std::map<std::thread::id,PerThread> pt;
```
instead, and return `pt[std::this_thread::get_id()]` instead. This also creates a global thread-local variable, but the compiler knows how to unify these variables across translation units.

You probably want to use a `static std::mutex` to access this variable since access to elements with a new thread-id triggers insertion of the value and may not be thread-safe on other threads.

**Comment by tjhei** [comment]
> Maybe we should look at (c) for now?

This would be great as taskflow is not usable in the current state for us. Are you planning on developing/intregrating c) and maintaining 2.x? Do you want help in form of PRs? If yes, I would need to know in more details where you want to go (always make a shared lib, provide a #define for header-only, etc.).

Another point I am curious about:
Did you consider the pros and cons of your switch to c++17 for 3.x and what this means for part of your user base?

**Comment by tjhei** [comment]
@bangerth This sounds more or less what I meant by option b).

**Comment by tsung-wei-huang** [comment]
@tjhei, yes, if you would be able to help create a pull request for enabling a shared library in the CMake, that's the best (I guess it won't be too complicated). This will also reduce the compilation time of taskflow, even though it's not large for now.

@bangerth  Having a std::map that hashes the given thread id into a worker id is valid but it can incur some cost because the executor frequently calls this per-thread id information to schedule tasks. I may not prefer to do this.

We are switching to C++17 because it largely simplifies our codebase and improves performance. The `inline variable` introduced in C++17 is another consideration to solve this static initialization order issue. There are a lot of new features that we rely on new C++17 STL and features which would otherwise be very difficult to do in C++14. See the [tentative release notes](https://taskflow.github.io/taskflow/release-3-0-0.html) for more details. Also, v3 will add a lot of new support for GPU programming on top of CUDA graph (though it requires CUDA v11.1 with c++17). 

At the same time, I will still provide support for the latest 2.7 if there are any critical issues. 
 
Just curious, what compiler version are you using now?

**Comment by tjhei** [comment]
thanks for the detailed infos!

> Just curious, what compiler version are you using now?

GNU/Linux: GCC version 5.0 or later; Clang version 4.0 or later; ICC versions 15 or later
Windows: MSVC 2017 or newer.

**Comment by bangerth** [comment]
I made an experiment in https://github.com/dealii/dealii/commit/103e5ae6a5e40d2f46d151f1858732c0a6538caa . Without this patch, @tjhei's testcase fails. With the patch, the testcase succeeds. So it really is the use of the `thread_local` variable.

I'm using GCC 7.5. I have to admit that I don't quite know what the standard really suggests happen -- the fact that we get different variables in different translation units would suggest that compilers treat `thread_local` variables as `static` *at the scope in which the function is defined* as opposed to a `static` local variable. I don't think that was the intent of the authors of the standard, but it seems to be what compiler writers implement. (I double-checked with writing `static thread_local`, and that has the same outcome.)

Regardless, this seems like a bug in TaskFlow then (possibly caused by a compiler bug). The commit linked to above fixes it. I understand that you might be concerned about a performance penalty, but in the end if you have the choice between a program aborting and a program possibly being a bit slower but correct, I'd still err for the latter :-)

**Comment by bangerth** [comment]
As for C++17: We can't switch to this yet, and probably can't for a few years to come. If you move TaskFlow to C++17, then we'll have to stick with version 2.x for the foreseeable future.

**Comment by tsung-wei-huang** [comment]
@bangerth @tjhei let me work on this and see if I can create a patch 2.7.1 for this. The static mutex will work because a static variable in an inline function guarantees to have one identity across transition units. Ideally, thread_local should also be implying static but I suspect the issue comes from GCC.

**Comment by bangerth** [comment]
I agree that this looks like a bug in the compiler. It simply doesn't make any sense that `thread_local` would imply "different across translation units" :-(

**Comment by tjhei** [comment]
I can confirm that my test case passes with gcc-10 and fails with gcc-7.5 (both in c++14 mode). Both pass with @bangerth 's workaround.

**Comment by tjhei** [comment]
I found this gcc bug that might be it: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90562

**Comment by tsung-wei-huang** [comment]
@tjhei  Yes, I believe this is a gcc bug when `static` goes together with `thread_local`. According to C++ standards, a local static variable within an inline function must guarantee a single view across transition units. 

I have tested gcc v8, v9, and v10 on Ubuntu and CentOS and didn't see this problem. Apple clang 8, 9, 10, and clang++ 6, 7, 8, 9, 10 are all ok on my platform. With gcc v7, the problem happens, and I will need to either (1) explicitly make `per_thread` an inline variable using `-std=c++17` or (2) add another layer of accessing static variable to force per-thread go through this guard so the compiler can be forced to see the static layer first.

I am thinking if there is a better way to do this because `per_thread` is in the **performance-critical** part. Each task execution goes through this `per_thread` to decide the worker that runs it in the work-stealing loop.

**Comment by bangerth** [comment]
@tjhei -- ah, good sleuthing. I had looked earlier today in the GCC bug database but hadn't made it to the CLOSED bug yet. I think that makes it pretty clear that it is indeed a bug.

@tsung-wei-huang I tried that with explicitly saying `static thread_local` but it made no difference. That's because `thread_local` implicitly makes the variable `static`. As a consequence, it makes no difference with regard to the compiler bug. If you follow the bug report @tjhei found to other links, you'll find that the bug is fixed at some point in the 8.x series and later, but not in the 7.x series. That's consistent with your observations.

The way pretty much every package works around these sorts of issues is via
```
  #if __GCC__ && (GCC_VERSION < 84000)   // or whatever the right preprocessor defines are
    // do the mutex thing here
    static std::map<std::thread::id,PerThread> x;
    return x[std::this_thread::get_id()];
  #else
    thread_local PerThread x;
    return x;
  #endif
```
If TaskFlow had a configuration system, one would write a small test that checks for the existence of the bug without trying to figure out version numbers -- but this should work and will have the desired performance unless you have a compiler that is broken -- which at some point in the future will mean that nobody will incur the penalty.

**Comment by tsung-wei-huang** [comment]
@bangerth  yes, I think this macro workaround is the best solution we may have now. I will create a path later to add this to 2.7.1. However, please keep in mind this is not the best solution and it can still break given that this is an intrinsic bug in GCC 7.x. For example, many Taskflow users would use a different compiler to compile GPU code and then use GCC to link objects together. The host compiler to run GPU can be different from the final one or others. In this case, the macro will still break and result in undefined behaviors I believe.

**Comment by bangerth** [comment]
Thanks, @tsung-wei-huang , for being willing to write up a patch.

I think it's worth considering the cost implications of the `std::map`, but I'm quite sure that you'll never be able to measure it. You're working with `O(100)` threads, so finding an element in a map requires <10 comparisons and following pointers. That's quite a low cost compared to what you will likely do with the result, namely set up and start a task. There is also a cost associated with getting the mutex, but the likelihood for contention of the mutex is very very small given that what you're doing under the mutex is so cheap -- and locking a mutex that is not contended is essentially for free. 

In other words, not wanting to always use a `std::map` to make the bug go away, and instead trying to come up with some complicated `#ifdef` scheme, seems like a premature optimization for no real benefit.

**Comment by tsung-wei-huang** [comment]
I assume this issue is solved with the adoption of a newer GCC that fixes the bug. Feel free to reopen it for any further discussion.

**Comment by foolnotion** [comment]
Hi, I believe I have the same or similar issue, with gcc 11.2:

```
/nix/store/l9s100fykjyqafvf4qlqfcabxv0m058b-taskflow-3.2.0/include/taskflow/core/executor.hpp:933: void tf::Executor::_invoke_dynamic_task_external(tf::Node*, tf::Graph&, bool): Assertion `worker && worker->_executor == this' failed.
```
This happens when my library is compiled as a shared library. I have a function `void Foo(tf::executor& executor, ...)` which is defined in a `cpp` file. Moving the function definition into the header or building the library as static makes the error go away.

**Comment by tsung-wei-huang** [comment]
@foolnotion would you please try the master branch again? There has been a lot of improvement over this problem. Specifically, `thread_local` has been removed and replaced with a lightweight `std::thread::id` mapping. I hope the present master branch fixes this issue caused by shared-library build.

**Comment by foolnotion** [comment]
@tsung-wei-huang thanks a lot, indeed it seems that the present code fixes this issue. I can confirm that my shared library works now with taskflow rev e46376dd294f355f1d3468292f709fa7beb9471c

**Comment by tsung-wei-huang** [comment]
@foolnotion thanks. This will be included to the next release.  

@tjhei @bangerth I am sorry for this late fix but it does take us quite a while to optimize the performance impact on this. Now the `thread_local` has been removed and replaced with an executor-specific efficient hash mapping (local, non-static), the problem should be fixed.  

Please reopen it if you see any issue.
